### PR TITLE
mem-ruby,configs: Add GPU GLC Atomic Resource Constraints

### PR DESCRIPTION
**Pull request by DanKouch**
Added a resource constraint, AtomicALUOperation, to GLC atomics performed in the TCC.

The resource constraint uses a new class, ALUFreeList array. The class assumes the following:
  - There are a fixed number of atomic ALU pipelines
  - While a new cache line can be processed in each pipeline each cycle, if a cache line is currently going through a pipeline, it can't be processed again until it's finished

Two configuration parameters have been used to tune this behavior:
  - tcc-num-atomic-alus corresponds to the number of atomic ALU pipelines
  - atomic-alu-latency corresponds to the latency of atomic ALU pipelines

Change-Id: I25bdde7dafc3877590bb6536efdf57b8c540a939

### CODE REVIEW

**Review by mattsinc — DISMISSED** [review]
This looks good to me, but @abmerop should check too before we checkin.

**Comment by powerjg** [comment]
I think it would be a good idea for @tiagormk (or he can point to someone else if I've tagged the wrong person). I know that there are others working on "far atomic" implementations in Ruby. It would be good if there was a just one implementation in the `ruby/structures` directory instead of two different implementations that both have to be maintained.

Since this implementation is "first", the question is whether CHI's far atomics can use this implementation (or use this implementation with minor changes). If the two implementations are very different, we may need to have a longer discussion.

Let's hold off on merging until we hear from the CHI folks.

**Comment by mattsinc** [comment]
> I think it would be a good idea for @tiagormk (or he can point to someone else if I've tagged the wrong person). I know that there are others working on "far atomic" implementations in Ruby. It would be good if there was a just one implementation in the `ruby/structures` directory instead of two different implementations that both have to be maintained.
> 
> Since this implementation is "first", the question is whether CHI's far atomics can use this implementation (or use this implementation with minor changes). If the two implementations are very different, we may need to have a longer discussion.
> 
> Let's hold off on merging until we hear from the CHI folks.

We met with those folks at your request, and the sense I got from the meeting was that the far atomics implementation was too different for there to be a single implementation that works for both -- basically their implementation would not easily work with vectors of requests like this one has to.  So we went ahead with pushing this one.  @tiagormk can of course chime in to the contrary though.

**Comment by powerjg** [comment]
Ok, that sounds good. I'd like to take a pass through the new Ruby code before merging (assuming I can find the time this week).

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc:30]: Missing copyright

**Review by powerjg — CHANGES_REQUESTED** [review]
With a quick glance, I notice that some files are missing copyright. I'll take a more detailed pass later this week.

**Comment by tiagormk** [comment]
> > I think it would be a good idea for @tiagormk (or he can point to someone else if I've tagged the wrong person). I know that there are others working on "far atomic" implementations in Ruby. It would be good if there was a just one implementation in the `ruby/structures` directory instead of two different implementations that both have to be maintained.
> > Since this implementation is "first", the question is whether CHI's far atomics can use this implementation (or use this implementation with minor changes). If the two implementations are very different, we may need to have a longer discussion.
> > Let's hold off on merging until we hear from the CHI folks.
> 
> We met with those folks at your request, and the sense I got from the meeting was that the far atomics implementation was too different for there to be a single implementation that works for both -- basically their implementation would not easily work with vectors of requests like this one has to. So we went ahead with pushing this one. @tiagormk can of course chime in the to contrary though.

My understanding is this change only adds resource modeling to the existing atomics in the viper protocol, so I don't see any blocker in merging this.

In parallel we can keep discussing potential for modeling multi-atomic requests in CHI so support the GPU case. Unless @mattsinc and @DanKouch have already decide to go ahead and add far atomics to viper.

**Comment by mattsinc** [comment]
> > > I think it would be a good idea for @tiagormk (or he can point to someone else if I've tagged the wrong person). I know that there are others working on "far atomic" implementations in Ruby. It would be good if there was a just one implementation in the `ruby/structures` directory instead of two different implementations that both have to be maintained.
> > > Since this implementation is "first", the question is whether CHI's far atomics can use this implementation (or use this implementation with minor changes). If the two implementations are very different, we may need to have a longer discussion.
> > > Let's hold off on merging until we hear from the CHI folks.
> > 
> > 
> > We met with those folks at your request, and the sense I got from the meeting was that the far atomics implementation was too different for there to be a single implementation that works for both -- basically their implementation would not easily work with vectors of requests like this one has to. So we went ahead with pushing this one. @tiagormk can of course chime in the to contrary though.
> 
> My understanding is this change only adds resource modeling to the existing atomics in the viper protocol, so I don't see any blocker in merging this.
> 
> In parallel we can keep discussing potential for modeling multi-atomic requests in CHI so support the GPU case. Unless @mattsinc and @DanKouch have already decide to go ahead and add far atomics to viper.

We did not add far atomics directly into VIPER.  However the support for our GLC atomics and your far atomics are similar (ignoring the vector part).

File: configs/ruby/GPU_VIPER.py
Comment (powerjg) [inline configs/ruby/GPU_VIPER.py:508]: Why is this default so different from the one in RubyCache.py?

File: src/mem/ruby/structures/RubyCache.py
Comment (powerjg) [inline src/mem/ruby/structures/RubyCache.py]: It would be helpful to add a comment here that these two parameters only apply to GPU atomics.

File: configs/ruby/GPU_VIPER.py
Comment (powerjg) [inline configs/ruby/GPU_VIPER.py:503]: Why is this default different from the default in RubyCache.py?

File: configs/ruby/GPU_VIPER.py
Comment (powerjg) [inline configs/ruby/GPU_VIPER.py:279]: Are we guaranteed that if these options are provided that each TCC `resourceStalls` will be true?

I.e., would it be possible for the user to specify `--tcc_num_atomic_alus=8` on the command line and then it be ignored by the simulation?

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.hh]: What is `idx` in this case? Are the ALUs used for atomic tagged based on the set index?

I think that you don't want to have any parameter here and that you want the ALUs to be, in some sense, fully associative. I could be wrong here, though.

Alternatively, if you want to model that there is 1 ALU per cache bank, you should probably pass the full address in and have the notion of bank mapping in this object.

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.hh]: Ditto above about `idx`

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.hh:30]: Needs a license statement, too.

**Review by powerjg — CHANGES_REQUESTED** [review]
A few more comments/questions after a more careful review :)

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (mattsinc) [reply src/mem/ruby/structures/ALUFreeListArray.hh]: @DanKouch can chime in, but ultimately I believe the idx is representing a bank of the L2 (or a pseudo-bank).  I thought some of the other code in this file was doing the bank mapping -- Dan can correct me if I'm wrong here.  Yes we want the ALUs to be fully associative, but because we need to ensure a given cache line waits the full time while the ALU is allowed to be reused for other cache lines to mimic the pipeline behavior of real GPUs.  So there will need to be some parameter, but we can certainly debate what the parameter is.

File: configs/ruby/GPU_VIPER.py
Comment (mattsinc) [reply configs/ruby/GPU_VIPER.py:279]: I am not aware of this being possible, but I'm also not sure what proof you would want either?

File: configs/ruby/GPU_VIPER.py
Comment (mattsinc) [reply configs/ruby/GPU_VIPER.py:503]: Our idea was that while the GPU currently only uses these parameters, we wanted to set GPU specific parameters/values in VIPER and not set the same values by default for anyone else trying to use these parameters for other code.  But if desired we can certainly make them the same given that no one else uses it as is.

Comment (powerjg) [reply configs/ruby/GPU_VIPER.py:503]: Whatever makes sense to reduce user confusion is good for me. Usually having multiple places with different defaults adds to confusion.

File: configs/ruby/GPU_VIPER.py
Comment (powerjg) [reply configs/ruby/GPU_VIPER.py:279]: The code should probably look something like

```
if <user specifies `--tcc_num_atomic_alus`>:
    for cache in <tcc_sim_object_list>:
        cache.resource_stalls = True
```

Without this, I think you'll find many people misusing this code (i.e., thinking that their command line parameter changes something when it doesn't).

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (powerjg) [reply src/mem/ruby/structures/ALUFreeListArray.hh]: Right now, the `idx` that is passed into this function is the set number. I don't think that is the intended behavior. It should probably be the bank number, if that's the intention. The other option, which may be easier, would be to just pass in the address and do all of the bank and associativity logic in this object.

**Comment by mattsinc** [comment]
@DanKouch just making sure you've seen the above comments/questions/requests?

**Comment by BobbyRBruce** [comment]
Is this still actively being worked on? I don't see any movement since July 25th.

**Comment by mattsinc** [comment]
yes

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.cc:30]: Added, thanks!

File: configs/ruby/GPU_VIPER.py
Comment (DanKouch) [reply configs/ruby/GPU_VIPER.py:503]: I've updated the defaults in RubyCache.py to match

File: configs/ruby/GPU_VIPER.py
Comment (DanKouch) [reply configs/ruby/GPU_VIPER.py:508]: Updated RubyCache.py to match

File: configs/ruby/GPU_VIPER.py
Comment (DanKouch) [reply configs/ruby/GPU_VIPER.py:279]: It looks like this will happen if `--no-tcc-resource-stalls` is supplied. Since tcc-num-atomic-alus defaults to 64, doing what the above psuedocode specifies would make it so that the `--no-tcc-resource-stalls` is ignored unless `--tcc-num-atomic-alus=0` is also specified, which seems like it might be more confusing. I could have it override if tcc-num-atomic-alus isn't 64, but specifying `--no-tcc-resource-stalls` seems rather explicit.

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.hh]: The intent here is to keep track of what cache lines are currently in an atomic ALU pipeline. Since the ALUs are intended to be fully-assosciative, we don't need the bank number. In the latest push I've updated the class to take the full address, then determine the cache line from there.

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.hh]: Addressed above.

File: configs/ruby/GPU_VIPER.py
Comment (mattsinc) [reply configs/ruby/GPU_VIPER.py:279]: Personally I agree with @DanKouch here, but since @powerjg is concerned about this and is away right now, we'll probably have to wait for him to return to comment on this.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (mattsinc) [inline src/mem/ruby/structures/ALUFreeListArray.cc]: Year should probably be 2023

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (mattsinc) [inline src/mem/ruby/structures/ALUFreeListArray.hh]: 2023

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (mattsinc) [inline src/mem/ruby/structures/ALUFreeListArray.cc]: beign --> being

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (mattsinc) [inline src/mem/ruby/structures/ALUFreeListArray.hh]: we should probably call this numALUs instead

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (mattsinc) [reply src/mem/ruby/structures/ALUFreeListArray.hh]: See comment on updated patch

**Review by mattsinc — CHANGES_REQUESTED** [review]
Some of these things @powerjg will have to chime in on once he is back, since what @DanKouch did seems reasonable to me but is contrary to what @powerjg asked for.  Beside that, there are a few minor changes to make in the requested changes below.  Thanks for fixing things @DanKouch !

**Review by mattsinc — DISMISSED** [review]
This meets the requests I had, so I'll approve for now to kick off tests.  But since @powerjg has a different idea of how the resource stall/alu dependence should be handled, we should probably wait until he is back (unless @BobbyRBruce disagrees).

**Comment by mattsinc** [comment]
@DanKouch I also see this at the bottom:

This branch is out-of-date with the base branch
Merge the latest changes from develop into this branch. 

If you see the same thing, you'll need to rebase before this could be merged in.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc]: This is likely to behave in unexpected ways. What happens if the cache (or controller) has a different clock period than the "ruby system"? I.e., you cannot assume all caches are in the same clock domain.

I suggest taking a `Tick` as the `accessLatency` and converting from cycles to ticks in the controller.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc]: I hate to block a change on this, but the code style is not being followed in this file.

Only class member variables should be `camelCase` and all local variables (including function parameters) should be `snake_case`

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc]: Missing a space after parens
```suggestion
         (accessQueue.back().startTick < oldestValidRecordStart)) {
```

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc]: Why can't you just use `accessQueue.size()`?

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc:66]: I'm not convinced this logic actually implements the behavior you're going for. I don't understand why things stay in the queue for some amount of time, but you only check for things in the queue at exactly `curTick`.

Could you describe the behavior you're trying to simulate in this function? With this knowledge, I may be able to help you implement what you're going for.

File: src/mem/ruby/structures/ALUFreeListArray.hh
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.hh]: Should be camelCase and should not have `m` at the beginning. 

Well, actually, I don't think a `RubySystem*` should be needed at all.

File: src/mem/ruby/structures/CacheMemory.cc
Comment (powerjg) [inline src/mem/ruby/structures/CacheMemory.cc]: Here you should be able to use a function that's something like `cyclesToTicks(p.atomicLatency)`

File: src/mem/ruby/structures/RubyCache.py
Comment (powerjg) [inline src/mem/ruby/structures/RubyCache.py:49]: What does 0 latency mean, and why is that the default?

**Review by powerjg — CHANGES_REQUESTED** [review]
Thanks for the updates. I think this is getting closer to the model we want to have, but there's still some room for improvements :)

**Comment by mattsinc** [comment]
@DanKouch just making sure you saw @powerjg 's comments?

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (abmerop) [reply src/mem/ruby/structures/ALUFreeListArray.cc]: I think the Tick thing makes sense too.

File: src/mem/ruby/structures/RubyCache.py
Comment (abmerop) [reply src/mem/ruby/structures/RubyCache.py:49]: This may be what the behavior was prior to the patch (assume no latency / latency is part of the enqueue latency for the slicc action)

**Review by abmerop — COMMENTED** [review]
This looks fine to me, but should address Jason's style comments.

File: src/mem/ruby/structures/RubyCache.py
Comment (powerjg) [reply src/mem/ruby/structures/RubyCache.py:49]: Ok. That information should be in a comment, then. Users will want to know what "0" means.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.cc:66]: The idea is that there are `numALUs` ALU pipelines that are `accessLatency` cycles deep. A new request can enter each pipeline each cycle, which means that each cycle `numALUs` requests can be added (hence why I'm checking for things at the queue in curTick). `accessLatency` is only being used to make sure that a request to a cache line being processed in a pipeline isn't started until the previous request to that cache line makes it all the way through (which is being done on lines 84-86).

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.cc]: @powerjg Would the clock period not correspond to the ruby_system passed to CacheMemory (which is wrapped as RubyCache) on say, [line 158 of MOESI_AMD_BASE.py](https://github.com/gem5/gem5/blob/develop/configs/ruby/MOESI_AMD_Base.py#L158)? If it doesn't, would BankedArray, which is used by dataArray and tagArray [also seems to be using this ruby_system](https://github.com/gem5/gem5/blob/develop/src/mem/ruby/structures/BankedArray.cc#L98) to get the clock period in ticks also need to be changed?

File: configs/ruby/GPU_VIPER.py
Comment (DanKouch) [reply configs/ruby/GPU_VIPER.py:279]: > It looks like this will happen if `--no-tcc-resource-stalls` is supplied. Since tcc-num-atomic-alus defaults to 64, doing what the above psuedocode specifies would make it so that the `--no-tcc-resource-stalls` is ignored unless `--tcc-num-atomic-alus=0` is also specified, which seems like it might be more confusing. I could have it override if tcc-num-atomic-alus isn't 64, but specifying `--no-tcc-resource-stalls` seems rather explicit.

@powerjg Are you ok with leaving it as-is in light of this?

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.cc]: I've updated ALUFreeListArray to take Ticks, rather than Cycles.

File: src/mem/ruby/structures/RubyCache.py
Comment (DanKouch) [reply src/mem/ruby/structures/RubyCache.py:49]: Added a comment in the latest patchset.

**Comment by DanKouch** [comment]
I believe I've addressed all comments in my latest patch set. I've also rebased.

**Comment by abmerop** [comment]
The changes in commit [af9beea](https://github.com/gem5/gem5/commit/af9beea6e607a21af0817f2b71bea0a2f3c007e1) look good to me.  Seems like something went wrong with the rebase though

File: configs/ruby/GPU_VIPER.py
Comment (powerjg) [reply configs/ruby/GPU_VIPER.py:279]: If the default is that you will have resource stalls, then this is OK. But if the default is that resources stalls are off, then this is very confusing for users other than the person who wrote the code.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [inline src/mem/ruby/structures/ALUFreeListArray.cc:82]: Shouldn't this line be `record.startTick <= curTick()`?

Otherwise, you are only checking for *a single tick* not reserving the ALU for the ALU's latency.

Furthermore, the next line is comparing this to the number of things in the queue, which will *always* be `false`. There is no way `accesses_this_tick` could be more than the number of records in `accessQueue`.

Generally, this is difficult to understand code because it's not written in a way that is directly understandable from the purpose. It would be better to have an array of N entries (where N is the number of ALUs). When you reserve and entry, you put the time that the entry will be free in the location. When you are checking for an available entry, you go through and find if there are any entries with `free time < curTick()`. 

I could be misunderstanding this code, but I would encourage you to write the code to be more like the intended behavior. E.g., if you want to "reserve an ALU for `latency` time, then write code that does that instead of using a circular buffer and counts the number of units in use.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [reply src/mem/ruby/structures/ALUFreeListArray.cc:66]: I'm going to resolve this conversation, but I *still* think the code is not doing what you intend. See my comment below.

File: src/mem/ruby/structures/RubyCache.py
Comment (powerjg) [reply src/mem/ruby/structures/RubyCache.py:49]: Thanks!

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.cc:82]: @powerjg 
The intent of the code isn't to reserve entire ALUs. As I'm assuming these ALUs are pipelined, multiple operations can be processed in a single ALU at once. _accessQueue doesn't keep track of ALUs_, but rather which cache lines are currently being processed.

The intent is rather to:
1. Restrict the number of ALU operations that can start in a cycle to _numALUs_
    - This is being accomplished in lines 71 to 74 (However, I was wrong in switching numALUs to accessQueue.size(), I'll switch that back as soon as I can)
2. Make sure that no ALU operation to a line already being processed can start until the previous operation has completed
    - This is accomplished in lines 77 to 79

This should model a group of numALUs pipelined ALUs that are accessLatency ticks deep, where a new operation can begin in each ALU every cycle.

Now that I think about it, however, I'd need to double-check that resource constraints for the same cycle are checked on the same tick. If they aren't I would need to change `curTick` to some sort of `curCycle`. I'm pretty sure this is the case, but I'll check again.

Comment (powerjg) [reply src/mem/ruby/structures/ALUFreeListArray.cc:82]: I see... I misunderstood what you were trying to model. My bad :).

It would be good to update the comments to make this more clear.

File: configs/ruby/GPU_VIPER.py
Comment (DanKouch) [reply configs/ruby/GPU_VIPER.py:279]: Yes, the default is to have resource stalls.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (DanKouch) [reply src/mem/ruby/structures/ALUFreeListArray.cc:82]: @powerjg I've added additional comments and made the correction I mentioned.

**Comment by DanKouch** [comment]
@powerjg @mattsinc @abmerop As #367 and now #397 address underlying issues with atomics, I've marked this as a draft PR for the time being. I don't think it should be merged in until those go in.

**Review by mattsinc — DISMISSED** [review]
From my perspective this is ok, although the open comments should probably be resolved by @powerjg since they reflect his concerns.

File: src/mem/ruby/structures/ALUFreeListArray.cc
Comment (powerjg) [reply src/mem/ruby/structures/ALUFreeListArray.cc:82]: Thanks for the updates!

**Review by powerjg — DISMISSED** [review]
lgtm. thanks for the updates!

**Comment by mattsinc** [comment]
@DanKouch it looks like there is a merge conflict in this patch, based on the above:

This branch has conflicts that must be resolved
Use the [web editor](https://github.com/gem5/gem5/pull/120/conflicts) or the to resolve conflicts.
Conflicting files
configs/ruby/GPU_VIPER.py

Please resolve this soon so we can push it in :)

**Comment by mattsinc** [comment]
@DanKouch : follow-up: the merge conflict is from #389 by @v-ramadas

**Comment by mattsinc** [comment]
@DanKouch : now that #397 is merged in, this is next.  It looks like we need to rebase to head of develop and resolve some merge conflicts though.  Let me know if you need help with this.  I'd like this to be merged in too with v23.1 since #397 is (and both are part of a larger fix).

**Review by mattsinc — APPROVED** [review]
As far as I can tell, this just rebased the previous commit and resolved according merge issues, so carrying forward approval.

**Review by abmerop — APPROVED** [review]

**Comment by mattsinc** [comment]
@abmerop : any objections to merging this in?

**Comment by abmerop** [comment]
None

**Pull request merged**
